### PR TITLE
Adapt for latest http-client

### PR DESCRIPTION
**Pull request by danog**

### CODE REVIEW

**Comment by danog** [comment]
Ping @trowski @kelunik? :)

**Comment by nicolas-grekas** [comment]
Note after reviewing this PR: it would have been great to use PSR-17 instead of hardcoding against guzzle's PSR-7 implementation.

**Comment by kelunik** [comment]
@nicolas-grekas PR welcome! A deeper review is anyway pending before tagging.

**Comment by trowski** [comment]
@nicolas-grekas Reviewing the overall API of this library is on my TODO list.

Does `PsrAdapter` + `PsrHttpClient` not use PSR-17 as you'd expect? I have yet to try actually using this library in a project, so there may be some gaps in usability that I didn't notice on my recent cleanup pass.

File: src/AmpHandler.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        if (!\interface_exists(PromiseInterface::class)) {
+            throw new AssertionError("Please require guzzle to use the guzzle AmpHandler!");
+        }
+        $this->client = $client ?? (
+            (new HttpClientBuilder)
+            ->followRedirects(0)
+            ->build()
+        );
+        self::$psrAdapter ??= new PsrAdapter(new class implements RequestFactoryInterface {
```
Comment (nicolas-grekas) [inline src/AmpHandler.php:57]: this is the line that hardcodes guzzle's PSR-7

File: src/AmpHandler.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        }
+        $deferred = new DeferredCancellation;
+        $cancellation = $deferred->getCancellation();
+        $future = async(function () use ($request, $options, $cancellation) {
+            if (isset($options['delay'])) {
+                delay($options['delay'] / 1000.0, cancellation: $cancellation);
+            }
+            /** @psalm-suppress PossiblyNullReference Initialized in the constructor */
+            $request = self::$psrAdapter->fromPsrRequest($request);
+            if (isset($options[RequestOptions::TIMEOUT])) {
```
Comment (nicolas-grekas) [inline src/AmpHandler.php:83]: also this: this is not portable

**Comment by nicolas-grekas** [comment]
But maybe I misunderstood and this is just about having a guzzle adapter instead of a more generic PSR-foo adapter.

**Comment by trowski** [comment]
@nicolas-grekas Much of this PR was adding a Guzzle handler class, which maybe should have been done in a separate PR as it wasn't necessary for compatibility with amphp/http-client v5.x.

This library provides `PsrHttpClient`, which is a PSR-18 client using an instance of `PsrAdaptor` which uses PSR-17 factories to translate between PSR-7 requests/responses to Amp requests/responses.

[examples/psr-with-amp.php](https://github.com/amphp/http-client-psr7/blob/1.x/examples/psr-with-amp.php)

Is there something in the implementation I'm missing which makes this non-portable?

**Pull request merged**